ADR 59 evidence composition and lifecycle (#5331)
tendermint/tendermint: docs/architecture/adr-059-evidence-composition-and-lifecycle.md

```go
type Evidence interface { //existing
  Height() int64                                     // height of the offense
  Time() time.Time                                   // time of the offense
  Address() []byte                                   // address of the offending validator
  Bytes() []byte                                     // bytes which comprise the evidence
  Hash() []byte                                      // hash of the evidence
  Verify(chainID string, pubKey crypto.PubKey) error // verify the evidence
  Equal(Evidence) bool                               // check equality of evidence

  ValidateBasic() error
  String() string
}
```

```go
type DuplicateVoteEvidence struct {
  VoteA *Vote
  VoteB *Vote

  timestamp time.Time // taken from the block time
}
```

> Tendermint has now introduced a new type of evidence to protect light clients from being attacked. This `LightClientAttackEvidence` is vastly different to `DuplicateVoteEvidence` in that it is physically a much different size containing a complete signed header and validator set. It is formed within the light client, not the consensus engine and requires a lot more information from state to verify (`VerifyLightClientAttack(commonHeader, trustedHeader *SignedHeader, commonVals *ValidatorSet)`  vs `VerifyDuplicateVote(chainID string, pubKey PubKey)`). Finally it batches validators together as opposed to having individual evidence. This evidence stretches the existing mould that was used to accomodate new types of evidence and has thus caused us to reconsider how evidence should be formatted and processed.

Author: I can't find the spelling fix

> Tendermint has now introduced a new type of evidence to protect light clients from being attacked. This `LightClientAttackEvidence` is vastly different to `DuplicateVoteEvidence` in that it is physically a much different size containing a complete signed header and validator set. It is formed within the light client, not the consensus engine and requires a lot more information from state to verify (`VerifyLightClientAttack(commonHeader, trustedHeader *SignedHeader, commonVals *ValidatorSet)`  vs `VerifyDuplicateVote(chainID string, pubKey PubKey)`). Finally it batches validators together as opposed to having individual evidence. This evidence stretches the existing mould that was used to accomodate new types of evidence and has thus caused us to reconsider how evidence should be formatted and processed.

Author: ah accommodate

> Tendermint has now introduced a new type of evidence to protect light clients from being attacked. This `LightClientAttackEvidence` is vastly different to `DuplicateVoteEvidence` in that it is physically a much different size containing a complete signed header and validator set. It is formed within the light client, not the consensus engine and requires a lot more information from state to verify (`VerifyLightClientAttack(commonHeader, trustedHeader *SignedHeader, commonVals *ValidatorSet)`  vs `VerifyDuplicateVote(chainID string, pubKey PubKey)`). Finally it batches validators together as opposed to having individual evidence. This evidence stretches the existing mould that was used to accomodate new types of evidence and has thus caused us to reconsider how evidence should be formatted and processed.

Reviewer: Heh, sorry, probably could have made that clearer.

```go
type LightClientAttackEvidence struct { // proposed struct in spec
  ConflictingBlock *LightBlock

>   CommonHeight int64

Reviewer: `CommonHeight` is the last height in common between the... witness and provider?

Author: The witness and the primary. Both of which are providers

Reviewer: Ah, got it, thanks. Maybe: 

```suggestion
  CommonHeight int64 // the last height at which the primary provider and the witness provider agree
```

  Type  AttackType     // enum: {Lunatic|Equivocation|Amnesia}

  timestamp time.Time // taken from the block time at the common height
}

```

## Possible Approaches for Evidence Composition

### Individual framework

> Evidence remains on a per validator basis. This causes the least disruption to the current processes but requires that we break `LightClientAttackEvidence` into several pieces of evidence for each malicious validator. This not only has performance consequences in that there are n times as many database operations and that the gossiping of evidence will require more bandwidth then necessary (by requiring a header for each piece) but it potentially impacts our ability to validate it. In batch form, the full node can run the same process the light client did to see that 1/3 validating power was present in both the common block and the conflicting block whereas this becomes more difficult to verify individually without opening the possibility that malicious validators forge evidence against innocent . Not only that, but `LightClientAttackEvidence` also deals with amnesia attacks which unfortunately have the characteristic where we know the set of validators involved but not the subset that were actually malicious (more to be said about this later). And finally splitting the evidence into individual pieces makes it difficult

Reviewer: I think you dropped off mid-sentence here.

However individual evidence has the advantage that it is easy to check if a node already has that evidence meaning we just need to check hashes to know that we've already verified this evidence before. Batching evidence would imply that each node may have a different combination of duplicate votes which may complicate things.

#### An example of a possible implementation path

`LightClientAttackEvidence` won't change but the evidence interface will need to look like the proposed one above and `DuplicateVoteEvidence` will need to change to encompass multiple double votes. A problem with batch evidence is that it needs to be unique to avoid people from submitting different permutations.

## Decision

Use a hybrid design

We allow individual and batch evidence to coexist together, meaning that verification is done depending on the evidence type and that  the bulk of the work is done in the evidence pool itself (including forming the evidence to be sent to the application).

## Detailed Design

Evidence has the following simple interface:

```go
type Evidence interface {  //proposed
  Height() int64                                     // height of the offense

>   Bytes() []byte                                     // bytes which comprise the evidence

Reviewer: This is just a byte slice (i.e., it is unstructured) because the different kinds of evidence would have different structures. Is that right?

Author: Yup it's the bytes of the evidence when you marshal to proto. It was there before so I have left it there.

  Hash() []byte                                      // hash of the evidence
  ValidateBasic() error
  String() string
}
```

> The changing of the interface is therefore breaking to the block.

Reviewer: ```suggestion
The changing of the interface is therefore block-breaking. This change means that evidence is hashed differently, and so when evidence is included in a block (more on that process later), the hash of blocks changes as well. 

This means that this change would require a network upgrade/restart. Fortunately, we have one upcoming for Stargate, so it's no problem to include this work in the Tendermint Stargate release.
```

We have two concrete types of evidence that fulfil this interface

```go
type LightClientAttackEvidence struct {
  ConflictingBlock *LightBlock
  CommonHeight int64
}
```

> where the `Hash()` is the hash of the header and commonHeight

Reviewer: I would also describe (at a high level) how `ValidateBasic` and `Bytes` are implemented for these two types (`LightClientAttackEvidence` and `DuplicateVoteEvidence`).

```go
type DuplicateVoteEvidence {
  VoteA *Vote
  VoteB *Vote
}
```
where the `Hash()` is the hash of the two votes

> The first is generated in the light client, the second in consensus. Both are sent to the evidence pool through `AddEvidence(ev Evidence) error` where it first runs `Has(ev Evidence)` to check if it has already received it (by comparing hashes) and then  `Verify(ev Evidence) error` before adding it to the database. There are two databases: one

Reviewer: ```suggestion
The first is generated in the light client, the second in the consensus reactor. Both are sent to the evidence pool through `AddEvidence(ev Evidence) error` where it first runs `Has(ev Evidence)` to check if it has already received it (by comparing hashes) and then `Verify(ev Evidence) error` before adding it to the database. There are two databases: one
```

> for pending evidence that is not yet committed and another of the committed evidence (on the chain)

Reviewer: The code in the evidence package often refers to "stores" (e.g., the `evidenceStore` or `stateStore`). Are you talking about these "stores" when you talk about "databases" in this ADR? And if so, do these "databases" already exist in the current implementation and what are they called?

- Fetch the validator set and confirm that the address is in the set at the height of the attack

- Check that the chain ID and signature is valid.

For `LightClientAttack`

- Fetch the common signed header and val set from the common height and use skipping verification to verify the conflicting header

- Fetch the trusted signed header at the same height as the conflicting header and compare with the conflicting header to work out which type of attack it is and in doing so return the malicious validators.

  - If equivocation the validators that signed for the commits of both the trusted and signed header

  - If lunatic the validators from the common val set that signed in the conflicting block

  - If amnesia none (we can't know which validators are malicious). This also means that we don't
  send amnesia evidence to the application

- For each validator, check the look up table to make sure there already isn't evidence against this validator

After verification we persist the evidence with the following key `height/hash` to the pending evidence database in the evidence pool with the following format

```go

> type EvidenceInfo struct {

Reviewer: This is kind of a bike shed, but I think it would be helpful to give this a more descriptive name. The things that differentiate `EvidenceInfo` from `Evidence` are that... 

- `Evidence` hasn't been validated, but `EvidenceInfo` has
- `EvidenceInfo` can be persisted to disk and gossiped (right? or is `Evidence` what is gossiped, and then every node recreates their own `EvidenceInfo`?) 
- `EvidenceInfo` contains the original `Evidence` in it 
- `EvidenceInfo` has information on which validators were involved in the attack, and `Evidence` does not 
- `Evidence` can be created by a light client (at least, when it is `LightClientAttackEvidence`) but `EvidenceInfo` can only be created by a full node 

Is all that right?

Author: `EvidenceInfo` is `Evidence` plus some cached data

Author: `Evidence` is what is gossiped because it is smaller in size and nodes will have to validate things anyway

  ev Evidence
  time time.Time
  validators []Validator
  totalVotingPower int64
}
```

We persist more information than just the evidence as this will be useful in forming the ABCI Evidence that we will come to later

> Receiving broadcasts from other evidence reactors works in the same manner.

Reviewer: This sentence could mean a lot of different things, and I'm not sure which one it is - does this mean that evidence reactors gossip to one another directly, and then repeat the same verification/Evidence-to-evidenceInfo-conversion/persistence-to-db process independently?

> When it comes to prevoting and precomitting a proposal that contains evidence, the full node will once again

Reviewer: This paragraph feels like the introduction of a new "step" of this whole evidence lifecycle. Is this what happens when a proposer node decides it's time to propose a block that contains evidence? Or is this what happens when a validator node receives a proposal for a block with evidence in it and has to decide how to vote?

call upon the evidence pool to verify the evidence using `CheckEvidence(ev []Evidence)`:

> This loops through all the evidence first checking that nothing has been duplicated then:

Reviewer: This section (from this line to line 190) is hard to follow. It might be easier to follow if it were written out in a somewhat more linear fashion, and if you were a bit more rigorous about being clear about what "this" and "it" refer to, throughout. 

It could begin like this:

`CheckEvidence` verifies the evidence by:

1. Looping through all the evidence, checking that nothing has been duplicated
2. Running `fastCheck(ev Evidence)` on each `Evidence` item received. (`fastCheck` works similar to `Has` in that it can quickly check if this node has seen and verified this evidence before, but it differs from `Has` in that it works for `LightClientAttackEvidence.`) If `fastCheck` doesn't pass, this node hasn't seen the evidence before, and so this node will have to verify the evidence itself. 

And so on.

runs `fastCheck(ev evidence)` which works similar to `Has` but with `LightClientAttackEvidence` if it has the
same hash it then checks to see that the validators it has are all signers in the commit of the conflicting header. If it doesn't pass fast check (because it hasn't seen the evidence before) then it will have to verify the evidence.

runs `Verify(ev Evidence)` - Note: this also saves the evidence to the db as mentioned before.

The final part of the lifecycle is when the block is committed and the `BlockExecutor` tries to apply state. Instead of trying to form the information necessary to produce the ABCI evidence it gets the evpool to create the evidence in the update step as follows.

> `Update(Block, State) []abci.Evidence`

Reviewer: These two blocks of code also need a little more exposition.

```go
abciResponses.BeginBlock.ByzantineValidators = evpool.Update(block, state)
```

> Update does the following:

Reviewer: This section (on what `Update` does) is also hard to read, thanks in part to the formatting.

- increments state which keeps track of both the current time and height used for measuring expiry
- marks evidence as committed and saves to db. This prevents validators from proposing committed evidence in the future
  Note: the db just saves the height and the hash. There is no need to save the entire committed evidence
- forms abci evidence:
```go
for _, val := range evInfo.Validators {
  abciEv = append(abciEv, &abci.Evidence{
    Type: evType,
    Validator: val,
    Height: evInfo.ev.Height(),
    Time: evInfo.time,
    TotalVotingPower: evInfo.totalVotingPower
  })
}
```
and compiles it all together (note for `DuplicateVoteEvidence` the validators array size is 1)
- removes expired evidence from both pending and committed

The abci evidence is then sent via the `BlockExecutor` to the application.

To summarize, we can see the lifecycle of evidence as such:

> ![evidence_lifecycle](../imgs/evidence_lifecycle.png)

Reviewer: This is a really great diagram! Did you make this?? (And how?) 

I especially love the nod to transit maps and the little train at the top. 

A few notes:

1. The "key" for the different types looks like it's part of the flow--pulling it a little farther away would make it easier to parse. 
2. Similarly, a little star that says "start here, when the light client notices a divergence" would be helpful. Or maybe two different stars: "start here for light client attacks" and "start here for duplicate votes"
3. What do the big white (empty) rectangles represent? 
4. Is there significance to the ordering of things inside the evidence pool?

Author: 3. They are just stations/platforms
4. Not really, but I tried to capture the flow of information

> Evidence is first detected and created in the light client and consensus engine. It is verified and stored as `EvidenceInfo` and gossiped to the evidence pools in other nodes. The consensus engine later communicates with the evidence pool to either retrieve evidence to be put into a block, or verify the evidence the consensus engine has retrieved in a block. Lastly when a block is added to the chain, the block executor sends the committed evidence back to the evidence pool so a pointer to the evidence can be stored in the evidence pool and it can update it's height and time. Finally, it turns the committed evidence into ABCI evidence and through the block executor passes the evidence to the application so the application can handle it.

Reviewer: This may be context that most reviewers would already have, but I'm still confused about what exactly the evidence pool is. I think this ADR would be even better with an extra paragraph on what the evidence pool does/what logic it has/what it handles and persists.

Author: I will try add more of an introductory paragraph at the start of the detailed design. Essentially the entire ADR is dedicated towards the evidence pool and the duties it performs.

## Status

> Accepted

Reviewer: Oh??

Author: Haha this is just preemptive

# ADR 059: Evidence Composition and Lifecycle

## Changelog

- 04/09/2020: Initial Draft (Unabridged)
- 07/09/2020: First Version

## Scope

> This document is designed to collate together and surface some predicaments involving evidence in Tendermint: both it's composition and lifecycle. It then aims to find a solution to these. The scope does not extend to the verification nor detection of certain types of evidence but concerns itself mainly with the general form of evidence and its lifecycle.

Reviewer: ```suggestion
This document is designed to collate together and surface some predicaments involving evidence in Tendermint: both its composition and lifecycle. It then aims to find a solution to these. The scope does not extend to the verification nor detection of certain types of evidence but concerns itself mainly with the general form of evidence and its lifecycle.
```

## Background

> For a long time `DuplicateVoteEvidence`, formed in the consensus engine, was the only evidence Tendermint had. It was produced whenever two votes from the same validator in the same round

Reviewer: ```suggestion
For a long time `DuplicateVoteEvidence`, formed in the consensus reactor, was the only evidence Tendermint had. It was produced whenever two votes from the same validator in the same round
```

I'm being more precise on "engine" vs. "reactor" just because Tendermint itself is often referred to as a "consensus engine."

  send amnesia evidence to the application

- For each validator, check the look up table to make sure there already isn't evidence against this validator

After verification we persist the evidence with the following key: `height/hash` to the pending evidence database in the evidence pool with the following format

```go
type EvidenceInfo struct {
  ev Evidence
  time time.Time
  validators []Validator
  totalVotingPower int64
}
```

`time`, `validators` and `totalVotingPower` are need to form the ABCI Evidence that we will come to later.

#### Broadcasting and receiving evidence

The evidence pool also runs a reactor that broadcasts the newly validated
evidence to all connected peers.

Receiving evidence from other evidence reactors works in the same manner as receiving evidence from the consensus engine or a light client.

#### Proposing evidence on the block

When it comes to prevoting and precomitting a proposal that contains evidence, the full node will once again
call upon the evidence pool to verify the evidence using `CheckEvidence(ev []Evidence)`:

This performs the following actions:

1. Loops through all the evidence to check that nothing has been duplicated

> 2. For each evidence, run `fastCheck(ev evidence)` which works similar to `Has` but instead for `LightClientAttackEvidence` if it has the

Reviewer: This section is now easier to understand but I'm still confused about what's going on with this sentence - is `fastCheck` checking to see if this node has verified the evidence already, or is it actually verifying the evidence again?

Author: It's checking to see if it has verified the evidence already. The reason why we can't use the `Has` function I mentioned previously is that the hash of the lightclientattackevidence (which is what is used for checking equality) is just the header hash and the common height. It doesn't include the hash of the validators that voted for the header as there could be numerous variants of this and we don't want nodes creating multiple variants of the same evidence. However the fact that we define the hash this way means that someone could create the same hash but only have one or two validators sign the commit (instead of 1/3) so we need to make sure that the validators we found during our initial verification are present in the commit of the conflicting block. 

Perhaps this is something that should also be documented

Reviewer: > Perhaps this is something that should also be documented

Is the answer to this question ever "no"? 😛 

I'm not sure that level of detail is quite needed in this document, but it would be good to make sure we have that captured in a code comment or godoc for `fastCheck`.

Author: > Is the answer to this question ever "no"? 😛

Haha but perhaps sometimes it should be. 

I have included a bit more of a description about the hash further up

same hash it then goes on to check that the validators it has are all signers in the commit of the conflicting header. If it doesn't pass fast check (because it hasn't seen the evidence before) then it will have to verify the evidence.

3. runs `Verify(ev Evidence)` - Note: this also saves the evidence to the db as mentioned before.

#### Updating application and pool

The final part of the lifecycle is when the block is committed and the `BlockExecutor` tries to apply state. Instead of trying to form the information necessary to produce the ABCI evidence itself, it gets the evidence pool to create it when it calls `Update(Block, State) []abci.Evidence`

Here is the excerpt in the `ApplyBlock()` where the evidence pool is updated.
```go
abciResponses.BeginBlock.ByzantineValidators = evpool.Update(block, state)
```

The `Update()` function does the following:

- Increments state which keeps track of both the current time and height used for measuring expiry

> - Marks evidence as committed and saves to db. This prevents validators from proposing committed evidence in the future

Reviewer: ```suggestion
- Marks evidence as committed and saves to db. This prevents validators from proposing committed evidence in the future.
```

As a heads up - markdown doesn't render newlines unless there are two of them back to back, or if you explicitly force them with a double space at the end of the line.

Author: I actually separated them so that it was easier to read in their unrendered form. You made an earlier comment about it and I thought it may referring to that. I also skip lines in other places that I used bullet points

  Note: the db just saves the height and the hash. There is no need to save the entire committed evidence

- Forms ABCI evidence:  (note for `DuplicateVoteEvidence` the validators array size is 1)
  ```go
  for _, val := range evInfo.Validators {
    abciEv = append(abciEv, &abci.Evidence{
      Type: evType,   // either DuplicateVote or LightClientAttack
      Validator: val,   // the offending validator (which includes the address, pubkey and power)
      Height: evInfo.ev.Height(),    // the height when the offense happened
      Time: evInfo.time,      // the time when the offense happened
      TotalVotingPower: evInfo.totalVotingPower   // the total voting power of the validator set
    })
  }

>   ```

Reviewer: OK, so just to check my own understanding:

This snippet constructs a slice of `abci.Evidence`s. If the evidence passed in was `DuplicateVoteEvidence`, then it will be of length one. Otherwise, it will be of length _n_, where _n_ is the number of misbehaving validators.

Author: yes exactly

- Removes expired evidence from both pending and committed

> The ABCI evidence is then sent via the `BlockExecutor` to the application.

Reviewer: The BlockExecutor makes an ABCI call? What function does it call? What are the types it passes across the ABCI boundary?

## Status

Accepted

## Consequences

<!-- > This section describes the consequences, after applying the decision. All consequences should be summarized here, not just the "positive" ones. -->

### Positive

- Evidence is better contained to the evidence pool / module
- LightClientAttack is kept together (easier for verification and bandwidth)
- Variations on commit sigs in LightClientAttack doesn't lead to multiple permutations and multiple evidence
- Address to evidence map prevents dos attacks of evidence by a single validator

### Negative

> - Breaking change to Evidence interface and thus evidence

Reviewer: This sentence doesn't parse for me.

However individual evidence has the advantage that it is easy to check if a node already has that evidence meaning we just need to check hashes to know that we've already verified this evidence before. Batching evidence would imply that each node may have a different combination of duplicate votes which may complicate things.

#### An example of a possible implementation path

`LightClientAttackEvidence` won't change but the evidence interface will need to look like the proposed one above and `DuplicateVoteEvidence` will need to change to encompass multiple double votes. A problem with batch evidence is that it needs to be unique to avoid people from submitting different permutations.

## Decision

> The decision is to adopt a hybrid design

Reviewer: ```suggestion
The decision is to adopt a hybrid design.
```

#### Verification

`Verify()` does the following:

- Use the hash to see if we already have this evidence in our committed database.

- Use the height to check if the evidence hasn't expired.

- If it has expired then use the height to find the block header and check if the time has also expired in which case we drop the evidence

- Then proceed with switch statement for each of the two evidence:

For `DuplicateVote`:

- Check that height, round, type and validator address are the same

- Check that the Block ID is different

- Check the look up table for addresses to make sure there already isn't evidence against this validator

- Fetch the validator set and confirm that the address is in the set at the height of the attack

- Check that the chain ID and signature is valid.

For `LightClientAttack`

- Fetch the common signed header and val set from the common height and use skipping verification to verify the conflicting header

- Fetch the trusted signed header at the same height as the conflicting header and compare with the conflicting header to work out which type of attack it is and in doing so return the malicious validators.

>   - If equivocation the validators that signed for the commits of both the trusted and signed header

Reviewer: ```suggestion
  - If equivocation, return the validators that signed for the commits of both the trusted and signed header
```

>   - If lunatic the validators from the common val set that signed in the conflicting block

Reviewer: ```suggestion
  - If lunatic, return the validators from the common val set that signed in the conflicting block
```

>   - If amnesia none (we can't know which validators are malicious). This also means that we don't

Reviewer: ```suggestion
  - If amnesia, return no validators (since we can't know which validators are malicious). This also means that we don't
```

>   send amnesia evidence to the application

Reviewer: ```suggestion
currently send amnesia evidence to the application, although we'll introduce more robust amnesia evidence handling in future Tendermint Core releases.
```

- For each validator, check the look up table to make sure there already isn't evidence against this validator

> After verification we persist the evidence with the following key: `height/hash` to the pending evidence database in the evidence pool with the following format

Reviewer: ```suggestion
After verification we persist the evidence with the key `height/hash` to the pending evidence database in the evidence pool with the following format:
```

Also, I really love the precision in this sentence around _what_ is getting written and _where_ it is being written!

```go
type EvidenceInfo struct {
  ev Evidence
  time time.Time
  validators []Validator
  totalVotingPower int64
}
```

> `time`, `validators` and `totalVotingPower` are need to form the ABCI Evidence that we will come to later.

Reviewer: ```suggestion
`time`, `validators` and `totalVotingPower` are need to form the `abci.Evidence` that we will send to the application layer. (More on this to come later.)
```

call upon the evidence pool to verify the evidence using `CheckEvidence(ev []Evidence)`:

This performs the following actions:

1. Loops through all the evidence to check that nothing has been duplicated

2. For each evidence, run `fastCheck(ev evidence)` which works similar to `Has` but instead for `LightClientAttackEvidence` if it has the
same hash it then goes on to check that the validators it has are all signers in the commit of the conflicting header. If it doesn't pass fast check (because it hasn't seen the evidence before) then it will have to verify the evidence.

3. runs `Verify(ev Evidence)` - Note: this also saves the evidence to the db as mentioned before.

#### Updating application and pool

The final part of the lifecycle is when the block is committed and the `BlockExecutor` then updates state. As part of this process, the `BlockExecutor` gets the evidence pool to create a simplified format for the evidence to be sent to the application. This happens in `ApplyBlock` where the executor calls `Update(Block, State) []abci.Evidence`.

```go
abciResponses.BeginBlock.ByzantineValidators = evpool.Update(block, state)
```

> Here is the format of the evidence that the application will receive. As seen above, this is stored as an array within `BeginBlock`.

Reviewer: 💯 , thank you for adding this!

> The changes to the application are minimal (it is still formed one for each malicious validator) with the exception of using an enum instead of a string for the evidence type.

Reviewer: ```suggestion
As far as the application is concerned, the changes introduced by Tendermint Core 0.34 are minimal. Previous versions of Tendermint also had an `abci.Evidence` type, and 0.34 continues to form one `Evidence` object for each malicious validator. The only change is that 0.34 uses an enum instead of a string for the `Type`.
```

(You don't need to accept this change; it's just a suggestion, trying to make it as clear as possible.)

## Status

Accepted

## Consequences

<!-- > This section describes the consequences, after applying the decision. All consequences should be summarized here, not just the "positive" ones. -->

### Positive

- Evidence is better contained to the evidence pool / module
- LightClientAttack is kept together (easier for verification and bandwidth)
- Variations on commit sigs in LightClientAttack doesn't lead to multiple permutations and multiple evidence

> - Address to evidence map prevents dos attacks of evidence by a single validator

Reviewer: ```suggestion
- Address to evidence map prevents DOS attacks, where a single validator could DOS the network by flooding it with evidence submissions
```

Evidence remains on a per validator basis. This causes the least disruption to the current processes but requires that we break `LightClientAttackEvidence` into several pieces of evidence for each malicious validator. This not only has performance consequences in that there are n times as many database operations and that the gossiping of evidence will require more bandwidth then necessary (by requiring a header for each piece) but it potentially impacts our ability to validate it. In batch form, the full node can run the same process the light client did to see that 1/3 validating power was present in both the common block and the conflicting block whereas this becomes more difficult to verify individually without opening the possibility that malicious validators forge evidence against innocent . Not only that, but `LightClientAttackEvidence` also deals with amnesia attacks which unfortunately have the characteristic where we know the set of validators involved but not the subset that were actually malicious (more to be said about this later). And finally splitting the evidence into individual pieces makes it difficult

#### An example of a possible implementation path

> We would ignore amnesia evidence (as individually it's hard to make) and revert to the initial split we had before where `DuplicateVoteEvidence` is also used for light client equivocation attacks and thus we only need `LunaticEvidence`. We would also most likely need to remove `Verify` from the interface as this isn't really something that can be used.

Reviewer: I'm confused about why Verify isn't useful anymore - can you say more? Is it because you can no longer verify every evidence type?

Author: It's just because the data required (i.e. the arguments needed) to have a generalised verify function is very broad. It's still possible if we wanted to but we would have to pass both the block store and state store as arguments which kind of raises the barriers of verifying say DuplicateVoteEvidence when for this particular evidence we just need a valset and a chainID

Reviewer: Ah, got it, hmm. It _feels_ like `Verify` is something that we should be able to call on all evidences. I wonder if there's another way to access the block store and state store without passing them in as parameters.

Author: In the spec it is something like `func IsValid(lcaEvidence LightClientAttackEvidence, bc Blockchain) boolean` in which they pass the entire blockchain

Reviewer: lol. _here, have a blockchain_

Last important note is that we are closest to this implementation than any of the other two.

#### An example of a possible implementation path

We would have to change the interface as it's not possible for `LightClientAttackEvidence` to return a single `Address()` so instead we would have `Addresses()`. In addition the actual infringing validators aren't known in `LightClientAttackEvidence` until verification is done as we are taking the union between the validator set at the common height and at the conflicting height.

```go
type Evidence interface { //proposed

>   Height() int64                                     // height of the equivocation

Reviewer: These comments all say "equivocation," but it's not necessarily actually equivocation, right?

Author: correct - I just plucked it out of the code but these will need to be updated

  Time() time.Time                                   // time of the equivocation
  Addresses() []Address                              // addresses of the equivocating validators
  Bytes() []byte                                     // bytes which comprise the evidence
  Hash() []byte                                      // hash of the evidence
  Equal(Evidence) bool                               // check equality of evidence

  ValidateBasic() error
  String() string
}
```

> Most likely we would therefore need to internally save the infringing validators which would make it easier to form abci evidence once the evidence has been committed on to chain. The `BlockExecutor` shouldn't have too much evidence specific logic in forming the abci evidence. Hence `LightClientAttackEvidence` should look like:

Reviewer: At this point, we don't actually know which of the "infringing validators" have actually been malicious, right? We won't know that until the evidence has been verified?

#### An example of a possible implementation path

`LightClientAttackEvidence` won't change but the evidence interface will need to look like the proposed one above and `DuplicateVoteEvidence` will need to change to encompass multiple double votes. A problem with batch evidence is that it needs to be unique to avoid people from submitting different permutations.

> I most likely have not captured the full picture of all these approaches and therefore would encourage others to add their takes on each of them.

## Evidence Interface

> *Note: any changes to the interface will be changing the block data structure and is hence breaking*

Reviewer: Block-breaking, but not ABCI breaking, right?

As already been alluded to it is most likely that we will need to remove `Verify()` as it's difficult to define a set of arguments that will satisfy the requirements for both duplicate vote verification and light client attack verification (not to mention future evidence and their requirements).

Depending on the composition we might need to change `Address()` to accomodate either individual and/or batch evidence. It may also be a good idea to have internal fields that the evidence pool populates such as time and validators so modules outside of the evidence pool (specifically the `BlockExecutor` don't need to query for more resources)

`Equal()` can also be seen as redundant if we set `Hash()` in the right way that we can just compare evidence hashes for uniqueness. We already use hashes for database lookups

`Type()` could be a new method which would be helpful in constructing abci evidence

> `ToABCI()` is another alternative that was written as a TODO in the actual code base.

Reviewer: What would this function do? Turn an `Evidence` into whatever type we pass over ABCI?

Author: Yup it was a note made in the codebase that I thought could be one approach. Actually what I finally propose is something similar where instead of the block executor creating the abci evidence the evidence pool does

## The State of Amnesia

TBC

## Censorship

A second problem brought up in a prior [issue](https://github.com/tendermint/tendermint/issues/4593) is the notion of evidence censorship. Consider the following scenario: I have greater than 1/3 validator power. I perform an attack either on a light client of on the main network that was detected. Whenever someone proposes a block that contains evidence I abstain from voting meaning that it never is committed. When it is my turn to propose, I propose a block that does not include the evidence, everyone else votes and that block gets committed. Eventually the chain reaches a height and time where the evidence has expired and any attempt to punish my attack has been thwarted.

This would most likely need an RFC (in order to explore possible solutions). As far as I can see there is a 0% approach a 50% approach and 100% approach in terms of the level of intervention.

The 0% approach would just be simply to leave this as is.

> The 50% approach would be that validators would consider a proposal invalid if it didn't contain at least one of the evidence that that validator had and thus wouldn't vote on it. This solution is far less intrusive to the current consensus algorithm but it also doesn't quite solve the problem of censorship, rather this protocol will keep the chain halted until off-chain governance can intervene.

Reviewer: I'm having trouble parsing the first sentence in this paragraph. Can you rephrase?

Author: Will do

Reviewer: The new phrasing makes perfect sense, thanks!

The 100% approach is a lot more intricate and would be to find a protocol where upon receiving a valid proposal with evidence, current validators that had been shown to have misbehaved would be barred from voting such that the total voting power would be reduced by the voting power of the malicious validators and this would allow for 2/3 consensus to be reached regardless of what the malicious validators did.

## Proposal

A Hybrid design

Evidence has the following simple interface:

```go
type Evidence interface {  //proposed
  Height() int64                                     // height of the equivocation
  Bytes() []byte                                     // bytes which comprise the evidence
  Hash() []byte                                      // hash of the evidence
  ValidateBasic() error
  String() string
}
```

> We have two recognized evidence that complies with it

Reviewer: "There are two concrete types that fulfill the `Evidence` interface." 

(Is that the same thing that you mean?)

Author: yup

```go
type LightClientAttackEvidence struct {
  ConflictingBlock *LightBlock
  CommonHeight int64
}
```
where the `Hash()` is the hash of the header and commonHeight

```go
type DuplicateVoteEvidence {
  VoteA *Vote
  VoteB *Vote
}
```
where the `Hash()` is the hash of the two votes

The first is generated in the light client, the second in consensus. Both are sent to the evidence pool through `AddEvidence(ev Evidence) error` where it first runs `Has(ev Evidence)` to check if it has already received it (by comparing hashes) and then  `Verify(ev Evidence) error` before adding it to the database

`Verify()` does the following:

> - Use the hash to see if we already have this evidence in our committed database.

Reviewer: Can you explain more about the "committed database"? When does evidence enter that database?

For more context on how I'm thinking about this, here are my notes on the first part of the light client evidence lifecycle:

1. Light client notices that something is amiss
2. Light client constructs a `LightClientAttackEvidence` object 
3. Light client passes this object to a full node--typically the full node that is serving as its "witness"
4. The full node verifies that this `LightClientAttackEvidence` is valid by calling `Verify`
5. If valid, the full node gossips the `LightClientAttackEvidence` to other full nodes. 
6. Once a proposing node has the `LightClientAttackEvidence` on hand, it will include it in the next block it proposes

Where in this process does the evidence enter a database?

Author: Between step 4 and 5. We persist data in case something happens to the evidence pool

Reviewer: OK. So from this and the description above, along with a peek at the actual code, is it fair to say that the evidence pool is almost like a reactor itself, in that it has both a bunch of business logic as well as ways to persist data (through databases it holds onto)? 

And the evidence reactor deals with evidence that comes in BOTH from the light client and from other full nodes (like when the evidence is being gossiped)?

Author: I just had this discussion with Marko about this. To me the reactor is this little socket thing on the side that interfaces with the p2p world but I think to most people the reactor encompasses the pool or whatever logic the rest of the module performs. So perhaps I should call everything the evidence reactor

Reviewer: "Reactor," yet another ambiguous and overloaded term in our codebase...

Author: Perhaps we need a `GLOSSARY.md`?

For `DuplicateVote`:

- Check that height, round, type and validator address are the same

- Check that the Block ID is different

- Check the look up table for addresses to make sure there already isn't evidence against this validator

- Fetch the validator set and confirm that the address is in the set

- Check that the chain ID and signature is valid.

For `LightClientAttack`

- Fetch the common signed header and val set from the common height and use skipping verification to verify the conflicting header

- Fetch the trusted signed header at the same height as the conflicting header and compare with the conflicting header to work out which type of attack it is and in doing so return the malicious validators.

  - If equivocation the validators that signed for the commits of both the trusted and signed header

  - If lunatic the validators from the common val set that signed in the conflicting block

  - If amnesia none (we can't know which validators are malicious). This also means that we don't
  send amnesia evidence to the application

- For each validator, check the look up table to make sure there already isn't evidence against this validator

> After verification we save the evidence with the following key `height/hash` and in the following struct

Reviewer: When you say "save," are you talking about persisting the evidence to the aforementioned database? 

(Clearly I'm just missing this database from my mental model of the evidence flow.)

Reviewer: ....is this the same thing as the evidence pool?

Author: yes for both statements

Reviewer: Wait, I'm thoroughly confused now - you said above that data is persisted to this db in case something happens to the evidence pool. Does the evidence pool persist data? Is it the only way that data is persisted?

```go
type EvidenceInfo struct {
  ev Evidence
  time time.Time
  validators []Validator
  totalVotingPower int64
}
```

> We save extra information as this will be useful in forming the ABCI Evidence that we will come to later

Reviewer: Does this mean that evidence effectively metamorphoses from `Evidence` (either `LightClientAttackEvidence` or `DuplicateVoteEvidence`) to `EvidenceInfo` to `ABCIEvidence` as it passes through this evidence flow?

Author: Yes Evidence is what is created during detection, gossiped and committed on chain. EvidenceInfo contains extra information and is stored in the evidence pool. ABCI Evidence is the evidence sent to the application

Reviewer: ![complete-metamorphosis](https://user-images.githubusercontent.com/773524/92262809-f4c4ad00-eedb-11ea-984c-978f0aefc792.gif)

Reviewer: (I feel we may want somewhat more descriptive names for these phases.)

Author: I like this but I don't think it is quite so linear nor process-like

Reviewer: Haha, ok. Is that because evidence can go back and forth between these various "phases"? Or is it because evidence objects can split and merge at various phases?

Author: It's like we use different data structures to represent the information that the parts need. The application only needs abci.Evidence (they don't need all the verification data), the light client only knows as much information as `LightClientAttackEvidence` which is still small to gossip and commit but large enough for the full nodes (evidence pool) to fill in the rest of the gaps by adding the information in `EvidenceInfo`

Reviewer: Sure. Are those transformations non-linear though?

(I just want my evidence to be a beautiful butterfly!)

Author: ![evidence_lifecycle](https://user-images.githubusercontent.com/27455974/92373062-1309f300-f0fe-11ea-8900-3aa75746f068.png)

Not as pretty as yours @tessr but how about this as a schematic

This loops through all the evidence first checking that nothing has been duplicated then:

runs `fastCheck(ev evidence)` which works similar to `Has` but with `LightClientAttackEvidence` if it has the
same hash it then checks to see that the validators it has are all signers in the commit of the conflicting header. If it doesn't pass fast check (because it hasn't seen the evidence before) then it will have to verify the evidence.

runs `Verify(ev Evidence)` - Note: this also saves the evidence to the db as mentioned before.

The final part of the lifecycle is when the block is committed and the `BlockExecutor` tries to apply state. Instead of trying to form the information necessary to produce the ABCI evidence it gets the evpool to create the evidence in the update step as follows.

`Update(Block, State) []abci.Evidence`

```go
abciResponses.BeginBlock.ByzantineValidators = evpool.Update(block, state)
```

Update does the following:
- increments state which keeps track of both the current time and height used for measuring expiry
- marks evidence as committed and saves to db. This prevents validators from proposing committed evidence in the future
- forms abci evidence:
```go

> for _, val := range evInfo.Validators

Reviewer: How does this `abci.Evidence` actually get "across" the ABCI boundary and into the application? Does the `BlockExecutor` make an ABCI call?

Author: yes

  Bytes() []byte                                     // bytes which comprise the evidence
  Hash() []byte                                      // hash of the evidence
  ValidateBasic() error
  String() string
}
```

We have two concrete types of evidence that fulfil this interface

```go
type LightClientAttackEvidence struct {
  ConflictingBlock *LightBlock
  CommonHeight int64
}
```
where the `Hash()` is the hash of the header and commonHeight

```go
type DuplicateVoteEvidence {
  VoteA *Vote
  VoteB *Vote
}
```
where the `Hash()` is the hash of the two votes

The first is generated in the light client, the second in consensus. Both are sent to the evidence pool through `AddEvidence(ev Evidence) error` where it first runs `Has(ev Evidence)` to check if it has already received it (by comparing hashes) and then  `Verify(ev Evidence) error` before adding it to the database. There are two databases: one
for pending evidence that is not yet committed and another of the committed evidence (on the chain)

`Verify()` does the following:

- Use the hash to see if we already have this evidence in our committed database.

- Use the height to check if the evidence hasn't expired.

> - If it hasn't expired use the height to find the block header and confirm that the time hasn't also expired.

Reviewer 2: Since evidence expiration is based on time & height, will the check take time from the lightblock?

> - If it hasn't expired use the height to find the block header and confirm that the time hasn't also expired.

Reviewer 2: It should be height && time not only height then only time. I believe there was a small issue when it was check if height expired and then time.

Author: yeah sorry in my head, I had it the right way around but I wrote it the other. It should be:
```
- If it has expired then use the height to find the block header to see if the time has also expired. If so then stop
```
basically replace hasn't with has

- Then proceed with switch statement for each of the two evidence:

For `DuplicateVote`:

- Check that height, round, type and validator address are the same

- Check that the Block ID is different

- Check the look up table for addresses to make sure there already isn't evidence against this validator

- Fetch the validator set and confirm that the address is in the set at the height of the attack

- Check that the chain ID and signature is valid.

For `LightClientAttack`

- Fetch the common signed header and val set from the common height and use skipping verification to verify the conflicting header

- Fetch the trusted signed header at the same height as the conflicting header and compare with the conflicting header to work out which type of attack it is and in doing so return the malicious validators.

  - If equivocation the validators that signed for the commits of both the trusted and signed header

  - If lunatic the validators from the common val set that signed in the conflicting block

  - If amnesia none (we can't know which validators are malicious). This also means that we don't
  send amnesia evidence to the application

> - For each validator, check the look up table to make sure there already isn't evidence against this validator

Reviewer 3: what if it's a newer evidence? old one might expire

Author: So my thinking on this is that we want to have this look up table as a way to stop malicious validators from spamming the network by creating thousands of evidence right? So if each evidence is actually valid and we always check and keep the newest one then this is not going solve our problem as we will keep having to validate and update our lookup table and then try gossip every new evidence that comes in. In all honesty, I would like to think that this evidence expiration will never really be a problem unless of course someone is intentionally censoring the network (I'm expecting there to be a large buffer between trusting period and expiration period).

However, if we wanted to try do both we could have a check to see if the existing header is in the later half of that period  and if so then we can look to swap it out with the newer evidence

Reviewer 3: yeah, let's implement a simpler version first

## Possible Approaches

### Individual framework

Evidence remains on a per validator basis. This causes the least disruption to the current processes but requires that we break `LightClientAttackEvidence` into several pieces of evidence for each malicious validator. This not only has performance consequences in that there are n times as many database operations and that the gossiping of evidence will require more bandwidth then necessary (by requiring a header for each piece) but it potentially impacts our ability to validate it. In batch form, the full node can run the same process the light client did to see that 1/3 validating power was present in both the common block and the conflicting block whereas this becomes more difficult to verify individually. Not only that, but `LightClientAttackEvidence` also deals with amnesia attacks which unfortunately have the characteristic where we know the set of validators involved but not the subset that were actually malicious (more to be said about this later).

> ### Hybrid Framework

Reviewer 3: another hybrid approach could be 

a) send / receive evidence in batches (duplicate votes, light client attacks, etc.)
b) break evidence per validator 
c) store also per validator & height

Author: and how would evidence be committed on chain? individual or batch

Reviewer 3: on evidence is in blocks, so it's already "batched" in this way.

Author: Sure but you could have two different data structures, a batch of individual evidence or a single batched evidence, where the latter is likely to be smaller in size

Reviewer 3: sure

We allow individual and batch evidence to coexist together, meaning that verification, broadcasting and committing is done depending on the evidence type. A hybrid model effectively undermines any possibility of generalization and Tendermint's evidence module will use switch statements and concrete types instead of interfaces to correctly process the evidence. One could argue that although we could see more evidence on the horizon, this won't be to the extent of justifying generalizing evidence and that each evidence can make it's own unique way through the pipeline and on to being committed to the chain and broadcasted to the application.

> A hybrid framework (as well as a batch framework) also surfaces the problem of the interface between Tendermint and the application. Currently, we have one concrete type that captures the relevant information needed to communicate to the application. This concrete type is formed per validator and therefore makes it more difficult to assess cartel (multi-validator) attacks. This begs the question of whether Tendermint should, at the application level, break down the evidence for each of the validators or alter the ABCI data structure to include multiple validators that were involved at that height and thus better capture the full extent of the misbehavior.

Reviewer 3: my sense is that ABCI application only needs two things: 1) address of the validator 2) type of misbehavior to understand how much to slash their stake

Author: perhaps also the power of the validator to assess the severity of the attack
i.e. duplicate vote of someone with 1 % power vs duplicate vote of someone with 34% power

Reviewer 3: abci app knows the power of each validator

Author: Okay but if you look at the abci evidence we send validator power and total validator power

Reviewer 3: oh, okay. not a big deal anyway

Last important note is that we are closest to this implementation than any of the other two.

### Batch Framework

> The last approach of this category would be to consider batch only evidence. This works fine with `LightClientAttackEvidence` but would require alterations to `DuplicateVoteEvidence` which would most likely mean that the consensus would send conflicting votes to a buffer in the evidence module which would then wrap all the votes together per height before gossiping them to other nodes and trying to commit it on chain. At a glance this may improve IO and verification speed and perhaps more importantly grouping validators gives the application and Tendermint a better overview of the severity of the attack. However individual evidence has the advantage that it is easy to check if a node already has that evidence meaning we just need to check hashes to know that we've already verified this evidence before. Batching evidence would imply that each node may have a different combination of duplicate votes which may complicate things.

Reviewer 3: > Batching evidence would imply that each node may have a different combination of duplicate votes which may complicate things.

exactly. or different alterations of LCEvidence: say one removes 1 vote and sends it. does it count as same evidence? what do you do in such case? or maybe previously it was incomplete (e.g. 20 signatures) and now you send complete (25 signatures) evidence

Reviewer 3: Also, malicious val can vote for 1000000 alternative blocks, which you will end up processing. But if evidence is based on validator, then you can reject other 9999999 evidences and only store 1, for which the validator will be punished!

Author: Yes I agree with you. I think the evidence module will need to have some address to evidence look up table to prevent this from happening

Author: As to the different combinations thing I think that is tricky
